framework/update/helper/swap.cpp

```
@@ -0,0 +1,265 @@
+    return true;
+}
+
+//! Atomically exchange `a` and `b` in a single syscall, so that there is no
+//! moment when the install location is empty or half-populated. Fails when the
+//! filesystem does not support it; the caller then falls back to two renames.
+bool exchangePaths(const fs::path& a, const fs::path& b)
+{
+#if defined(__APPLE__)
+    return ::renamex_np(a.c_str(), b.c_str(), RENAME_SWAP) == 0;
+#elif defined(SYS_renameat2)
+    return ::syscall(SYS_renameat2, AT_FDCWD, a.c_str(), AT_FDCWD, b.c_str(), RENAME_EXCHANGE) == 0;
```

Clang (14.0.6)

[warning] do not call c-style vararg functions

(cppcoreguidelines-pro-type-vararg)